src/hooks/sounds.ts

```diff
@@ -1,78 +1,97 @@
-import { useCallback } from "react";
-import { ContextRef } from "~/containers/sound-provider/index.js";
+import { useCallback, useEffect, useRef } from "react";
+import { type MatchStage } from "~/types/index.js";
+import { type TrackId } from "~/constants/sounds.js";
+import { SoundContextRef } from "~/containers/sound-provider/index.js";
 import { useAndRequireContext } from "~/hooks/use-and-require-context.js";
+import { useRouter } from "next/router.js";
 
-export const usePlayMusic = () => {
-  const { audioContext, masterGainNode, musicGainNode } = useAndRequireContext(
-    ContextRef,
-    "usePlayMusic",
-    "sound-provider",
-  );
-  return useCallback(
-    async (audioFile: string, loop?: boolean) => {
-      if (!audioContext || !masterGainNode || !musicGainNode) {
-        return;
-      }
-      try {
-        const response = await fetch(audioFile);
-        const arrayBuffer = await response.arrayBuffer();
-        await audioContext.decodeAudioData(
-          arrayBuffer,
-          (audioBuffer) => {
-            const source = audioContext.createBufferSource();
-            source.buffer = audioBuffer;
-            source.loop = !!loop;
-            source.connect(musicGainNode);
-            musicGainNode.connect(masterGainNode);
-            masterGainNode.connect(audioContext.destination);
-            source.start(0);
-          },
-          (e) => {
-            console.error("Error decoding audio:", e);
-          },
+export const usePlayMusic = (
+  audioFile: TrackId,
+  loop?: boolean,
+  pathname?: string,
+  delayInSeconds = 0,
+  stage?: MatchStage,
+  definedStage?: MatchStage,
+  offset?: number,
+  duration?: number,
+) => {
+  const { audioContext, audioBuffers, masterGainNode, musicGainNode } =
+    useAndRequireContext(SoundContextRef, "usePlayMusic", "sound-provider");
+
+  const sourceNodeRef = useRef<AudioBufferSourceNode | null>(null);
+  const router = useRouter();
+
+  useEffect(() => {
+    const playAudio = () => {
+      if (!audioContext || !masterGainNode || !musicGainNode) return;
+      const audioBuffer = audioBuffers.current.get(audioFile);
+      if (audioBuffer) {
+        const sourceNode = audioContext.createBufferSource();
+        sourceNode.buffer = audioBuffer;
+        sourceNode.loop = !!loop;
+        sourceNode.connect(musicGainNode);
+        musicGainNode.connect(masterGainNode);
+        masterGainNode.connect(audioContext.destination);
+        sourceNode.start(
+          audioContext.currentTime + delayInSeconds,
+          offset,
+          duration,
         );
-      } catch (error) {
-        console.error("Error playing sound: ", error);
+        sourceNodeRef.current = sourceNode;
       }
-    },
-    [audioContext, masterGainNode, musicGainNode],
-  );
+    };
+
+    // Function to stop the music
+    const stopAudio = () => {
+      if (sourceNodeRef.current) {
```

Comment:
Would a conditional chain be enough for this case? 
` sourceNodeRef.current?.stop()`